### ISSUE TITLE
Document the "upgrade my dependancies" workflow

### ISSUE THREAD

**Issue by jacobian** [pypa/pipenv#1490]
As far as I can tell, the documentation doesn't seem to lay out a proscribed workflow for upgrading dependencies. I think this is a fairly common use-case, something like this:

> As a user of pipenv, I've got a project with a `Pipfile` and a `Pipfile.lock`. It's been some time, and I'd like to know what's changed upstream, and to upgrade my dependencies (and test them as they go). 

I'm happy to contribute this documentation myself, and I think I've figured out a good workflow. But I'm opening this issue to make sure I'm on the write track before I write something.

Here's what I think the upgrade workflow should look like:

1. Find out what's changed upstream: `pipenv run pip list --outdated`
2. Upgrade packages, two options:
    a. Want to upgrade everything? Just do `pipenv install`.
    b. Want to upgrade packages one-at-a-time? `pipenv install --selective upgrade <pkg>` for each outdated package.
3. Run your tests, fix issues. If there are problems, pin versions in `Pipfile`.
4. Commit your upgraded `Pipfile.lock` (and perhaps `Pipfile`).

Does this sound right?

Something I don't understand about `--selective-upgrade` -- why does it modify `Pipfile` to add `keep_outdated`?

**Comment by kennethreitz** [comment]
@jacobian sounds like we have some work to do, regarding your last point there.

**Comment by kennethreitz** [comment]
That workflow sounds solid.

**Comment by kennethreitz** [comment]
I feel a pain point in `pip list --outdated`. I'd like to add a command that does that for you.

**Comment by jacobian** [comment]
I was wondering about that. 

How about `pipenv graph --outdated`? Unlike `pip list --outdated`, it could show where in the dep tree the outdated thing lives. I think this is important information: I probably want to upgrade top-level dependencies (i.e. the things that I depend on), but unless there's a security issue or a serious bug, I probably _don't_ want to bother upgrading my secondary dependencies until I do a top-level upgrade.

**Comment by xsteadfastx** [comment]
`pipenv install` doesnt upgrade all packages here in 10.1.0. so whats the workflow? so docs would be great :)

**Comment by kennethreitz** [comment]
that sounds like a great suggestion.

**Comment by ZhukovGreen** [comment]
Previously, when I was updating all packages, I was doing the following `pipenv update --dev`.
Now, `rm Pipfile.lock` and `pipenv install`.

**Comment by zyv** [comment]
Hi there!

Just upgraded to 10.1.0 and found out that `update` vanished. In the past, I would run `pipenv update --dry-run` to find out what packaged needed updating and then `pipenv update` to actually update them. I didn't quite gather from this issue why `update` got removed and what the recommended workflow is right now, but if I understand the last posts correctly, the alternative currently is to run

1. `pipenv run pip list --outdated`
2. `rm Pipfile.lock && pipenv install` or is `pipenv install --skip-lock` the way to go?

Unfortunately, this isn't equivalent to `pipenv update --dry-run`, because `pip list --outdated` doesn't consider the restrictions on package versions in the Pipfile, so, for example, it reports Django 2.0.2 as being available, whereas I pinned it to the latest LTS in the Pipfile.

Can we maybe get a `--dry-run` mode for `pipenv install --skip-lock` that would ignore the lockfile and tell what packages it is going to update?!

Also, I'm wondering why one option is called `--skip-lock` and the opposite of it (the default?!) is called `--ignore-pipfile`... I would put them next to each other in the docs and I guess one could reconsider the naming, e.g. could `--ignore-pipfile` && `--ignore-lock` or `--skip-lock` && `--skip-pipfile` or `--ignore-pipfile` and `--update` (or `--upgrade`) have been better names?

Thanks!

**Comment by TimSimpsonR** [comment]
>> Also, I'm wondering why one option is called --skip-lock and the opposite of it (the default?!) is called --ignore-pipfile

Actually, I believe the default for `pipenv install` is `--skip-lock`. In my experience `pipenv install` will change the lock file if it detects an old dependency, as will `pipenv sync`. The only way to get it to use the lock file without modifying it is `pipenv install --ignore-pipfile` (*note: I'm using pipenv version 10.1.0)*

I am having a hard time understanding why keeping the `update` command wasn't clearer than the current situation. I also believed -maybe wrongly- that the whole point of pipenv was it stuck to using the lockfile in most situations while making it obvious how to `update` it (pun intended), but now it seems like getting it to not touch the lockfile takes a lot of effort.

**Comment by egustavson** [comment]
Should `pipenv install` upgrade everything to latest with `version="*"`? I have seen that it is not doing this. I am using another source that points to an internal pypi.

**Comment by kakulukia** [comment]
please bring back `update`

**Comment by rhymes** [comment]
What about two new commands:

* `pipenv outdated`
* `pipenv update` (and pipenv update --dev) 

?

**Comment by kennethreitz** [comment]
``update`` makes no sense when we have ``sync``. 

There needs to be an outdated command, and it'll likely be a flag to an existing command.

I'd like to keep the list of commands as short as possible, you see.

**Comment by techalchemy** [comment]
So can we get a plan together? I like the idea in the initial suggestion if adding `—outdated` to `pipenv graph`.  

@kennethreitz to avoid confusion should we toss an update command in which only contains a warning to switch to sync until 10.2 or 11?

**Comment by kennethreitz** [comment]
no, it's too late. switch already happened. It was across a major version for a reason.

**Comment by kennethreitz** [comment]
We could add a vanity 'update' (which isn't listed in help), which prints out what to run now.

**Comment by kennethreitz** [comment]
e.g. `pipenv lock, pipenv sync`

**Comment by techalchemy** [comment]
Yeah that was my thinking. And exit with an error code so it still technically fails

**Comment by rgant** [comment]
Best I can tell the purpose of `sync` isn't even [documented](https://docs.pipenv.org/#pipenv-sync). I have no idea why I would want to sync. I don't know what it does. I want to install and upgrade my dependencies. Seems like that should be a first order use case to cover somehow.

@kennethreitz 
> `update` makes no sense when we have `sync`.

What do you mean by that? How are they equivalent or related?

**Comment by kennethreitz** [comment]
sync replaced update.

**Comment by kennethreitz** [comment]
@techalchemy should we add 'update' which runs lock then sync?

**Comment by rgant** [comment]
Replaced it how?

Ok, obviously this is in flux and you guys are busy sorting it out.

I don't think I'm ready for pipenv 10.1, I'll roll back and wait for there to be some specific documentation for how I figure out if there are updates I can apply and how to do that.

**Comment by techalchemy** [comment]
I am becoming persuaded of this use case

**Comment by kennethreitz** [comment]
this is why we have version numbers :)

**Comment by techalchemy** [comment]
That might be more intuitive... if I was coming at this as a brand new user who had no background in pip tools or anything I would probably just try to run `pipenv update`

**Comment by kennethreitz** [comment]
https://github.com/pypa/pipenv/commit/1d021cd3ffabb718c3ea43dc5b64e3567fec2e8c

**Comment by kakulukia** [comment]
Oh, up until now i thought that lock is locking my installed modules at their installed versions and writes the lock file, so i can keep those .. obviously i totally misunderstood that feature. Am i alone with that assumption?

**Comment by greysteil** [comment]
Might make sense to stop displaying `lock` as a user option - feels more useful (and designed for) automated dependency update tools?

**Comment by kennethreitz** [comment]
Okay so in mater, `update` is restored.

New workflow

<img width="679" alt="screen shot 2018-02-28 at 10-1 09 44 am" src="https://user-images.githubusercontent.com/119893/36794876-8a1a5342-1c6f-11e8-8150-a9702bdb0169.png">

**Comment by kennethreitz** [comment]
Providing a package name to `$ pipenv update` (e.g. `$ pipenv update requests`) actually runs `$ pipenv install requests --selective-upgrade`.

**Comment by kennethreitz** [comment]
Released as v0.10.1.

**Comment by kennethreitz** [comment]
@jacobian that should help with your workflow docs.

**Comment by kakulukia** [comment]
running pipenv update (probably even other commands) without a Pipfile in my directory, results in the following error:

    File ".../pipenv/project.py", line 251, in parsed_pipfile
    with open(self.pipfile_location) as f:
    TypeError: expected str, bytes or os.PathLike object, not NoneType

**Comment by kennethreitz** [comment]
gah, will fix. 

other commands don't have this issue.

**Comment by kennethreitz** [comment]
@kakulukia next time open a new issue instead of noising up @jacobian's issue please :)

Fixed, and released.

**Comment by kennethreitz** [comment]
## Recommended Workflow

- Find out what's changed upstream: `$ pipenv update --outdated`
- Upgrade packages, two options:
  a. Want to upgrade everything? Just do `$ pipenv update`.
  b. Want to upgrade packages one-at-a-time? `$ pipenv update <pkg>` for each outdated package.

Run your tests, fix issues. If there are problems, pin versions in `Pipfile`.

**Comment by vovkd** [comment]
What if i need install exactly locked dependecies, without updating Pipfile.lock
![image](https://user-images.githubusercontent.com/1261144/44704573-0772a300-aac6-11e8-9748-3a6e61462ff4.png)
How  to disable this behavior? Because i dont want automatically update all dependencies. The same is with installing with pipenv new package pipenv inistall some_package -- it will update all packages listed in Pipfile.lock

**Comment by rhymes** [comment]
@vovkd try with `pipenv install --keep-outdated`

**Comment by rhymes** [comment]
@vovkd yeah, my option didn't work. I like pipenv but I'm a little disappointed about this behavior. I'm sure it's very complicated to handle dependecies in general, hopefully in the future this will get better.

**Comment by timkofu** [comment]
Hi,

My dependencies are out of date and I aim to upgrade all of them in one go. So I started following the guidance here https://pipenv.readthedocs.io/en/latest/basics/#example-pipenv-upgrade-workflow

> pipenv update --outdated
✔ Success!
Skipped Update of Package wrapt: 1.10.11 installed, 1.11.1 available.
Skipped Update of Package whitenoise: 4.1 installed, 4.1.2 available.
Skipped Update of Package WebTest: 2.0.32 installed, 2.0.33 available.
Skipped Update of Package WebOb: 1.8.2 installed, 1.8.5 available.
Skipped Update of Package waitress: 1.1.0 installed, 1.2.1 available.
Skipped Update of Package uWSGI: 2.0.17.1 installed, 2.0.18 available.
Skipped Update of Package six: 1.11.0 installed, 1.12.0 available.
Skipped Update of Package redis: 2.10.6 installed, 3.1.0 available.
Skipped Update of Package pytz: 2018.5 installed, 2018.9 available.
Skipped Update of Package pylint: 2.1.1 installed, 2.2.2 available.
Skipped Update of Package psycopg2: 2.7.5 installed, 2.7.7 available.
Skipped Update of Package hiredis: 0.2.0 installed, 1.0.0 available.
Skipped Update of Package Django: 2.1.5 installed, 2.1.7 available.
Skipped Update of Package django-redis: 4.9.0 installed, 4.10.0 available.
Skipped Update of Package cffi: 1.11.5 installed, 1.12.0 available.
Skipped Update of Package beautifulsoup4: 4.6.3 installed, 4.7.1 available.
Skipped Update of Package astroid: 2.0.4 installed, 2.1.0 available.
All packages are up to date!

Second step:

> pipenv update .
Warning: . was not found in your Pipfile! Aborting.
Aborted!

And that's where it ends for me. What I'm I doing wrong?

shell: zsh
pipenv: version 2018.11.26
